src/clib/pio_msg.c

```diff
@@ -1802,295 +2372,433 @@ int PIOc_Init_Intercomm(int component_count, MPI_Comm peer_comm,
-	if (!my_iosys->comp_rank)
+	/* Is this task is part of a computation communicator? */
+	if (io_comm == MPI_COMM_NULL)
 	{
-	    if ((mpierr = MPI_Comm_rank(peer_comm, &iam)))
-		ierr = check_mpi(NULL, mpierr,__FILE__,__LINE__);
+	    /* Initialize the computational component. */
+	    ierr = init_comp_comm(comp_comms[cmp1], my_iosys, peer_comm, cmp1,
+				  &comp_leader, &io_leader);
 	}
 	else
-	    iam = -1;
-
-	/* Find the lucky comp_leader task. */
-	if ((mpierr = MPI_Allreduce(&iam, &comp_leader, 1, MPI_INT, MPI_MAX, peer_comm)))
-	    ierr = check_mpi(NULL, mpierr,__FILE__,__LINE__);
-
-	/* Is this the compmaster? Only if the comp_rank is zero. */
-	if (!my_iosys->comp_rank)
 	{
-	    my_iosys->compmaster = MPI_ROOT;
-	    comp_master = MPI_ROOT;
+	    /* Initialize the IO component. */
+	    ierr = init_io_comm(io_comm, my_iosys, peer_comm, cmp1,
+				&comp_leader, &io_leader);
 	}
-	else
-	    my_iosys->compmaster = MPI_PROC_NULL;
 
-	/* Set up the intercomm from the computation side. */
-	if ((mpierr = MPI_Intercomm_create(my_iosys->comp_comm, 0, peer_comm,
-					   io_leader, cmp, &my_iosys->intercomm)))
-	    ierr = check_mpi(NULL, mpierr,__FILE__,__LINE__);
+	/* Set up the union communicator, an intracomm that includes all
+	 * tasks in a computational component, plus all tasks in the IO
+	 * component. */
+	if (!ierr)
+	    ierr = init_union_comm(my_iosys);
+
+	/* Uncomment the following to print out some debug info. */
+	/* int my_rank; */
+	/* MPI_Comm_rank(MPI_COMM_WORLD, &my_rank); */
+	/* for (int t = 0; t < my_iosys->num_iotasks + my_iosys->num_comptasks; t++) */
+	/* { */
+	/*     MPI_Barrier(my_iosys->union_comm); */
+	/*     if (my_rank == t) */
+	/* 	pio_iosys_print(my_rank, my_iosys); */
+	/* } */
+
+	/* Add this id to the list of PIO iosystem ids. */
+	if (!ierr)
+	    iosysidp[cmp1] = pio_add_to_iosystem_list(my_iosys);
+	LOG((2, "added to iosystem_list iosysid = %d", iosysidp[cmp1]));
+    } /* next component */
 
-	/* Create the union communicator. */
-	if ((mpierr = MPI_Intercomm_merge(my_iosys->intercomm, 0, &my_iosys->union_comm)))
-	    ierr = check_mpi(NULL, mpierr,__FILE__,__LINE__);
-    }
-    else
+    /* Now call the function from which the IO tasks will not
+     * return until the PIO_MSG_EXIT message is sent. */
+    if (!ierr)
+	if (io_comm != MPI_COMM_NULL)
+	    if ((ierr = pio_msg_handler(my_iosys->io_rank, component_count, iosys)))
+		return ierr;
+
+    /* If there was an error, make sure all tasks see it. */
+    /* if (ierr) */
+    /* 	if ((mpierr = MPI_Bcast(&ierr, 1, MPI_INT, 0, iosys->intercomm))) */
+    /* 	    ierr = check_mpi(NULL, mpierr, __FILE__, __LINE__); */
+
+    return ierr;
+}
+
+/** Set up the communicators for parallel I/O.
+ *
+ * @param world the communicator containing all the available tasks.
+ * @param component_count number of computational components
+ * @param color for this task. Use 0 for an I/O task, 1 for the first
```

Comment:
I think that you are assuming each task is in only one computational group.  This is not correct.